source/registry/contracts/Registry.sol

```diff
@@ -338,16 +230,135 @@ contract Registry {
+    if (_transferAmount > 0) {
+      stakingToken.safeTransfer(msg.sender, _transferAmount);
+    }
+  }
+
+  /**
+   * @notice Return a list of all server URLs supporting a given token
+   * @param _token address of the token
+   * @return urls array of staker server URLs supporting the token
+   */
+  function getServerURLsForToken(
+    address _token
+  ) external view returns (string[] memory urls) {
+    EnumerableSet.AddressSet storage stakers = stakersByToken[_token];
+    uint256 _length = stakers.length();
+    urls = new string[](_length);
+    for (uint256 i = 0; i < _length; i++) {
+      urls[i] = stakerServerURLs[address(stakers.at(i))];
+    }
+  }
+
+  /**
+   * @notice Return whether a staker supports a given token
+   * @param _staker account address used to stake
+   * @param _token address of the token
+   * @return true if the staker supports the token
+   */
+  function supportsToken(
+    address _staker,
+    address _token
+  ) external view returns (bool) {
+    return tokensByStaker[_staker].contains(_token);
+  }
+
+  /**
+   * @notice Return a list of all supported tokens for a given staker
+   * @param _staker account address of the staker
+   * @return tokenList array of all the supported tokens
+   */
+  function getTokensForStaker(
+    address _staker
+  ) external view returns (address[] memory tokenList) {
+    EnumerableSet.AddressSet storage tokens = tokensByStaker[_staker];
+    uint256 _length = tokens.length();
+    tokenList = new address[](_length);
+    for (uint256 i = 0; i < _length; i++) {
+      tokenList[i] = tokens.at(i);
+    }
+  }
+
+  /**
+   * @notice Return a list of all stakers supporting a given token
+   * @param _token address of the token
+   * @return _stakers array of all stakers that support a given token
+   */
+  function getStakersForToken(
+    address _token
+  ) external view returns (address[] memory _stakers) {
+    EnumerableSet.AddressSet storage stakerList = stakersByToken[_token];
+    uint256 _length = stakerList.length();
+    _stakers = new address[](_length);
+    for (uint256 i = 0; i < _length; i++) {
+      _stakers[i] = stakerList.at(i);
+    }
+  }
+
+  /**
+   * @notice Get the ServerURLs for an array of stakers
+   * @param _stakers array of staker addresses
+   * @return _urls array of staker ServerURLs in the same order
+   */
+  function getServerURLsForStakers(
+    address[] calldata _stakers
+  ) external view returns (string[] memory _urls) {
+    uint256 stakersLength = _stakers.length;
+    _urls = new string[](stakersLength);
+    for (uint256 i = 0; i < stakersLength; i++) {
+      _urls[i] = stakerServerURLs[_stakers[i]];
+    }
+  }
+
   /**
    * @notice Return the staking balance of a given staker
    * @param _staker address of the account used to stake
    * @return balance of the staker account
    */
   function balanceOf(address _staker) external view returns (uint256) {
+    if (bytes(stakerServerURLs[_staker]).length == 0) return 0;
```

Comment:
Interesting that you can:
* addTokens
* addProtocols

But your balance remains zero until you set the URL. Also, until you set the URL, you cannot removeStakedServer, but you can removeProtocols and removeTokens.
I think balanceOf should return the staking balance whether you have set the URL or not.

Reply:
Since they're not useful without a URL, functions `addTokens` and `addProtocols` could require a URL to be set. Thoughts?

Reply:
I want to use balanceOf to simplify my system so it's most important for me that it is accurate. Of secondary concern is the cost of updating the token list. Of minimal concern is the cost of balanceOf, since I don't think it will be invoked much on-chain.